### ISSUE TITLE
send email update to all students

### ISSUE THREAD

**Issue by haraldschilly** [sagemathinc/smc#304]
add a facility to email all (or a selected few) students in a course.

**Comment by williamstein** [comment]
There is a serious subtle issue that we have to address soon.  Namely, if a spammer creates a course, they can then send spam to all "students" (=victims) in the course as us.  Inevitably this will happen.   When it does, we'll deal with it.   For now, doing what you describe here doesn't actually make things any worse, because a user can send other non-users arbitrary emails by inviting them to a course in the first place (which is already scary). 

In the long run what we need here is a way for a user to send a direct message to another user... or for a user to make a message associated with a project available.  Then users that have an email on file would get sent an email notification (configurable -- they can turn it off) telling them that this message is available (and maybe including the message).   Anyway, the point is that this facility for courses should be a special case of something a bit more general involving messaging.

**Comment by paulsmart** [comment]
How about make it a members-only feature?
It would discourage spammers and encourage membership.

**Comment by rbeezer** [comment]
See also  #274

**Comment by haraldschilly** [comment]
one way to deal with spammers is to classify all outgoing emails, flag and hold them back (internally) and then we review them. sendgrid has a feature to measure the spam-level, but no idea how to use that. also, the actual overhead is a management interface to deal with those messages which got flagged.

enabling this only for paying members is of course a good spam protection.

**Comment by williamstein** [comment]
Also requested by: Arnold Lebow

**Comment by williamstein** [comment]
Closely related to https://github.com/sagemathinc/smc/issues/700

**Comment by kedlaya** [comment]
Regarding the point that you can send someone an arbitrary email by inviting them to a course (or to be a collaborator on a project): this seems like it should also be true with other packages like Dropbox or Google Docs, and yet I have never gotten a spam invite from either. I would be very surprised if these were being screened manually.

**Comment by haraldschilly** [comment]
As an opposite example, I got spam invites for google calendar events. They
stopped after me and many complained about it. So yes, this happens, and
all those larger services have good mechanisms in place to handle this. We
can start using a spam filter for outgoing messages, but implementing and
managing that is not an easy task. Sendgrid offers some kind of outgoing
spam mail detections, but I don't know if that would be useful for our
purposes, nor do I know how good it really is.

Still, we definitely have to solve this, and we need a good plan how to do
this.

**Comment by kcrisman** [comment]
Yeah, any LMS has this capability, and not having it is a definite issue for practical usage.  Difference is that the Registrar (or whoever) signs up students, not you, so maybe spamming isn't as bad of an issue then.

**Comment by johnjeng** [comment]
So, an easy way to do this without breaking the current security model is let the instructor email everyone that they added by email. We already list the emails for which this is true **and** the links are `mail:to`'s. It makes sense for there to be a `mail all` button.
![emails](https://cloud.githubusercontent.com/assets/618575/22169742/71652e1c-df2b-11e6-85f5-98c5f5c49b19.png)

**Comment by williamstein** [comment]
> So, an easy way to do this without breaking the current security model is let the instructor email everyone that they added by email. 

+1 do this.  Also, add a feature to allow instructors to set the email address for a student.  We already have a field for this in the .course file, so that shouldn't be hard.

**Comment by rbeezer** [comment]
I was thinking about exactly this (and wondering how it might be a security 
hazard) when I had to alert the students to a mistake I'd made, and I had to 
descend down into our campus IT infrastructure, rather than the friendly 
confines of SMC.  +1, and thanks for working on this.

Rob

**Comment by johnjeng** [comment]
## Work so far

Default when there's no search.
<img width="780" alt="screen shot 2017-09-11 at 6 32 13 pm" src="https://user-images.githubusercontent.com/618575/30285910-06e867e0-9720-11e7-8774-aba7741d80fb.png">

Number omitted has been moved to below the search bar.
<img width="788" alt="screen shot 2017-09-11 at 6 29 22 pm" src="https://user-images.githubusercontent.com/618575/30285909-06df565a-9720-11e7-901a-62ad531cc32b.png">

Now the question is, should the button email all students displayed or always all students? I'm leaning towards all students displayed since it's easy enough to clear the search and maybe search has some sophisticated filtering abilities in the future.

I don't want to allow both options because that's option clutter in my opinion.

**Comment by kedlaya** [comment]
I could easily see myself accidentally emailing only a subset of my class when I meant to email the whole class, because I forgot to remove the filter. But that could be remedied by suitable labeling on whatever screen you reach by hitting the button.

**Comment by williamstein** [comment]
Can you make it “Email All Students...” with ..., since it will surely open up another dialog.

**Comment by haraldschilly** [comment]
once that's in, someone please remind me to look closely on the monthly quota for our emailing service :mag_right:

**Comment by williamstein** [comment]
REQUESTED BY: Brian Clowers

**Comment by mforbes** [comment]
Can this be enabled somehow or is it still in development?

**Comment by williamstein** [comment]
This is not implemented.

**Comment by novoselt** [comment]
What happened with the code that generated the screenshots? It seems that emailing to students added by an email was agreed upon.

**Comment by williamstein** [comment]
> What happened with the code that generated the screenshots?

Who knows -- John hasn't worked on cocalc for many years, so it's gone.  It'll have to be written starting from scratch.  Also, in 2017 the code was I think in a different language (coffeescript).

**Comment by novoselt** [comment]
> In the long run what we need here is a way for a user to send a direct message to another user...

That part is done now ;-) Presumably this issue becomes much easier now!

**Comment by williamstein** [comment]
>  Presumably this issue becomes much easier now!

It's easier but still significant work to do efficiently, unfortunately, because there is no notion of bcc or group in our messaging system.  **Sending a copy of a message individually to all students (rather than a big cc) would be easy but not efficient.**  Sending a single message to everybody without a huge list of recipients isn't implemented. 

I should probably just implement "Sending a copy of a message individually to all students ", since that would be very useful in most cases and doesn't require "groups" at all.
